src/apps/test/pfutest.cpp

```diff
@@ -0,0 +1,133 @@
+#include "pfu.hpp"
+#include "../test/MemFile.hpp"
+#include "../test/fileutils.hpp"
+#include "common/Configuration.hpp"
+#include "save.hpp"
+#include "string_utils.hpp"
+
+#include <catch2/catch.hpp>
+#include <fmt/format.h>
+
+extern void boot_db();
+extern void reap_old_chars();
+
+using namespace pfu;
+
+namespace {
+spdlog::logger logger = spdlog::logger("pfutest");
+}
+
+struct LoadTinyMudOnce : Catch::TestEventListenerBase {
+    using TestEventListenerBase::TestEventListenerBase;
+
+    void testRunStarting([[maybe_unused]] Catch::TestRunInfo const &testInfo) override {
+        setenv(MUD_AREA_DIR_ENV, TEST_DATA_DIR "/area", 1);
+        setenv(MUD_DATA_DIR_ENV, TEST_DATA_DIR "/data", 1);
+        setenv(MUD_HTML_DIR_ENV, TEST_DATA_DIR "/html", 1);
+        setenv(MUD_PORT_ENV, "9000", 1);
+        boot_db();
+    }
+    void testRunEnded([[maybe_unused]] Catch::TestRunStats const &testRunStats) override { reap_old_chars(); }
+};
+CATCH_REGISTER_LISTENER(LoadTinyMudOnce)
+
+TEST_CASE("collect player names") {
+    SECTION("all non-empty player files found") {
+        const auto player_dir = Configuration::singleton().player_dir();
+
+        const auto names = collect_player_names(player_dir, logger);
+
+        std::vector<std::string> expected = {"Versionfour"};
+        REQUIRE(names == expected);
+    }
+}
+
+TEST_CASE("register tasks") {
+    SECTION("expected task count") {
+        const auto all_tasks = register_tasks();
+
+        REQUIRE(all_tasks.size() == 1);
+    }
+}
+
+TEST_CASE("upgrade player") {
+    SECTION("v4 to latest") {
+        const auto all_tasks = register_tasks();
+        const auto player_dir = Configuration::singleton().player_dir();
+        test::MemFile god_file;
+        test::MemFile player_file;
+        const CharSaver saver;
+        auto result =
+            upgrade_player("Versionfour", all_tasks, logger, [&god_file, &player_file, &saver](const Char &ch) {
+                saver.save(ch, god_file.file(), player_file.file());
```

Comment:
Using the templated `upgrade_player()` to inject a saver that takes the output files so we can inspect them. Whereas `pfu` itself calls the single arg version of `save()`.

Reply:
Got it: I saw the template. I like it :)  I'd have probably "just" tested calling `cup.upgrade()` and then left the actual "save" part untested rather than add machinery to make it tested, but this is probably better :)